src/physics/sitetype.jl

```diff
@@ -224,22 +224,52 @@ function op(name::AbstractString, s::Index...; adjoint::Bool=false, kwargs...)
   # if there are multiple indices
   commontags_s = commontags(s...)
 
-  # XXX this clashes with the S+ and S- for spins...
-  ## Interpret operator names joined by +
-  ## as acting sequentially on the same site
-  #pluspos = findfirst("+", name)
-  #if !isnothing(pluspos)
-  #  op1 = name[1:prevind(name, pluspos.start)]
-  #  op2 = name[nextind(name, pluspos.start):end]
-  #  return op(op1, s...; kwargs...) + op(op2, s...; kwargs...)
-  #end
+  # first we handle the + and - algebra, which requires a space between ops to avoid clashing
+  name_split = nothing
+  @ignore_derivatives name_split = String.(split(name, " "))
+  oplocs = findall(x -> x ∈ ("+", "-"), name_split)
+
+  if !isempty(oplocs)
+    @ignore_derivatives !isempty(kwargs) &&
+      error("Lazy algebra on parametric gates not allowed")
+
+    # the string representation of algebra ops: ex ["+", "-", "+"]
+    labels = name_split[oplocs]
+    # assign coefficients to each term: ex [+1, -1, +1]
+    coeffs = [1, [(-1)^Int(label == "-") for label in labels]...]
+
+    # grad the name of each operator block separated by an algebra op, and do so by
+    # making sure blank spaces between opnames are kept when building the new block.
+    start, opnames = 0, String[]
+    for oploc in oplocs
+      finish = oploc
+      opnames = vcat(
+        opnames, [prod([name_split[k] * " " for k in (start + 1):(finish - 1)])]
+      )
+      start = oploc
+    end
+    opnames = vcat(
+      opnames, [prod([name_split[k] * " " for k in (start + 1):length(name_split)])]
+    )
 
-  # Interpret operator names joined by *
-  # as acting sequentially on the same site
-  starpos = findfirst("*", name)
-  if !isnothing(starpos)
-    op1 = name[1:prevind(name, starpos.start)]
-    op2 = name[nextind(name, starpos.start):end]
+    # build the vector of blocks and sum
+    op_list = [
+      coeff * (op(opname, s...; kwargs...)) for (coeff, opname) in zip(coeffs, opnames)
+    ]
+    return sum(op_list)
+  end
+
+  # the the multiplication come after
+  oploc = findfirst("*", name)
+  if !isnothing(oploc)
+    op1, op2 = nothing, nothing
+    @ignore_derivatives begin
+      op1 = name[1:prevind(name, oploc.start)]
+      op2 = name[nextind(name, oploc.start):end]
+      if !(op1[end] == ' ' && op2[1] == ' ')
+        @warn "composite op definition `A*B` deprecated: please use `A * B` instead (with spaces)"
```

Comment:
Let's use `Base.depwarn("composite op definition `A*B` deprecated: please use `A * B` instead (with spaces)", :op)`.

It is the function version of [`@deprecated`](https://docs.julialang.org/en/v1/base/base/#Base.@deprecate) which is the standard way in Julia to deprecate syntax, and can be turned on and off with a flag passed to `julia`.